File: src/codex/subscription.ts
import {
	requireValidOAuthCredentials,
	type StoredOAuthCredentials,
} from "../auth/credentials";
import { DEFAULT_CODEX_CLIENT_VERSION, resolveRelayUrl } from "./client";
import { ApiError, isAbortError } from "../shared/api-error";
import {
	isRecord,
	recordField,
	type JsonObject,
} from "../shared/json";
import {
	BodySizeLimitError,
	readLimitedBody,
} from "../shared/limited-body";

const CODEX_USAGE_PATH = "/backend-api/wham/usage";
const CODEX_USAGE_REQUEST_TIMEOUT_MS = 10_000;
const MAX_CODEX_USAGE_RESPONSE_BYTES = 256 * 1024;
const FIVE_HOUR_SECONDS = 5 * 60 * 60;
const WEEK_SECONDS = 7 * 24 * 60 * 60;
const MIN_MONTH_SECONDS = 28 * 24 * 60 * 60;
const MAX_MONTH_SECONDS = 31 * 24 * 60 * 60;

type SubscriptionEnv = Pick<
	Env,
	"AUTH_KV" | "DATA_ENCRYPTION_KEY" | "CODEX_RELAY_URL"
>;

export type CodexQuotaWindowKind =
	| "five_hour"
	| "weekly"
	| "monthly"
	| "primary"
	| "secondary";

export type CodexQuotaCategory = "codex" | "code_review" | "additional";

export interface CodexSubscriptionMetadata {
	planType: string | null;
	subscriptionActiveStart: number | null;
	subscriptionActiveUntil: number | null;
}

export interface CodexQuotaWindow {
	id: string;
	category: CodexQuotaCategory;
	name: string;
	kind: CodexQuotaWindowKind;
	usedPercent: number | null;
	remainingPercent: number | null;
	limitWindowSeconds: number | null;
	resetAt: number | null;
	allowed: boolean | null;
	limitReached: boolean;
}

export interface CodexRateLimitResetCredits {
	availableCount: number | null;
	applicableAvailableCount: number | null;
}

export interface CodexSubscriptionInfo extends CodexSubscriptionMetadata {
	windows: CodexQuotaWindow[];
	rateLimitResetCredits: CodexRateLimitResetCredits;
	fetchedAt: number;
}

export function codexSubscriptionMetadata(
	credentials: StoredOAuthCredentials,
): CodexSubscriptionMetadata {
	const claims = decodeJwt(credentials.idToken);
	const auth = recordField(claims, "https://api.openai.com/auth") ?? claims;
	return {
		planType: normalizePlanType(
			auth?.chatgpt_plan_type ??
				auth?.chatgptPlanType ??
				auth?.plan_type ??
				auth?.planType,
		),
		subscriptionActiveStart: dateValueToMs(
			auth?.chatgpt_subscription_active_start ??
				auth?.chatgptSubscriptionActiveStart ??
				auth?.subscription_active_start ??
				auth?.subscriptionActiveStart,
		),
		subscriptionActiveUntil: dateValueToMs(
			auth?.chatgpt_subscription_active_until ??
				auth?.chatgptSubscriptionActiveUntil ??
				auth?.subscription_active_until ??
				auth?.subscriptionActiveUntil,
		),
	};
}

export async function fetchCodexSubscription(
	env: SubscriptionEnv,
	clientSignal?: AbortSignal,
): Promise<CodexSubscriptionInfo> {
	const credentials = await requireValidOAuthCredentials(env);
	const response = await requestCodexUsage(env, credentials, clientSignal);
	const payload = await readUsagePayload(response, clientSignal);
	return codexSubscriptionFromUsage(
		payload,
		codexSubscriptionMetadata(credentials),
	);
}

export function codexSubscriptionFromUsage(
	payload: unknown,
	metadata: CodexSubscriptionMetadata,
	now = Date.now(),
): CodexSubscriptionInfo {
	if (!isRecord(payload)) throw invalidCodexUsageResponse();
	return {
		...metadata,
		planType:
			normalizePlanType(payload.plan_type ?? payload.planType) ??
			metadata.planType,
		windows: buildQuotaWindows(payload, now),
		rateLimitResetCredits: normalizeResetCredits(
			payload.rate_limit_reset_credits ?? payload.rateLimitResetCredits,
		),
		fetchedAt: now,
	};
}

async function requestCodexUsage(
	env: SubscriptionEnv,
	credentials: StoredOAuthCredentials,
	clientSignal: AbortSignal | undefined,
): Promise<Response> {
	const timeout = AbortSignal.timeout(CODEX_USAGE_REQUEST_TIMEOUT_MS);
	const signal = clientSignal
		? AbortSignal.any([clientSignal, timeout])
		: timeout;
	const headers = new Headers({
		Accept: "application/json",
		Authorization: `Bearer ${credentials.accessToken}`,
		"Content-Type": "application/json",
		"User-Agent": `codex_cli_rs/${DEFAULT_CODEX_CLIENT_VERSION}`,
	});
	if (credentials.accountId) {
		headers.set("Chatgpt-Account-Id", credentials.accountId);
	}

	let response: Response;
	try {
		response = await fetch(resolveUsageUrl(env.CODEX_RELAY_URL), {
			method: "GET",
			headers,
			redirect: "manual",
			signal,
		});
	} catch (error) {
		if (clientSignal?.aborted) throw clientAbortError(error);
		throw codexUsageUnavailable();
	}
	if (!response.ok) {
		await discardBody(response);
		throw codexUsageUpstreamError(response.status);
	}
	return response;
}

async function readUsagePayload(
	response: Response,
	clientSignal: AbortSignal | undefined,
): Promise<unknown> {
	let bytes: Uint8Array | null;
	try {
		bytes = await readLimitedBody(response, MAX_CODEX_USAGE_RESPONSE_BYTES);
	} catch (error) {
		if (clientSignal?.aborted) throw clientAbortError(error);
		if (error instanceof BodySizeLimitError) {
			throw invalidCodexUsageResponse();
		}
		throw codexUsageUnavailable();
	}
	if (!bytes) throw invalidCodexUsageResponse();
	try {
		return JSON.parse(new TextDecoder().decode(bytes)) as unknown;
	} catch {
		throw invalidCodexUsageResponse();
	}
}

function resolveUsageUrl(relayUrl: string): URL {
	const url = resolveRelayUrl(relayUrl);
	const suffix = "/backend-api/codex/responses";
	const normalizedPath = url.pathname.endsWith("/")
		? url.pathname.slice(0, -1)
		: url.pathname;
	const suffixIndex = normalizedPath.lastIndexOf(suffix);
	url.pathname =
		suffixIndex >= 0 && suffixIndex + suffix.length === normalizedPath.length
			? `${normalizedPath.slice(0, suffixIndex)}${CODEX_USAGE_PATH}`
			: CODEX_USAGE_PATH;
	url.search = "";
	return url;
}

function buildQuotaWindows(payload: JsonObject, now: number): CodexQuotaWindow[] {
	const windows: CodexQuotaWindow[] = [];
	appendRateLimitWindows(
		windows,
		payload.rate_limit ?? payload.rateLimit,
		{ id: "codex", category: "codex", name: "Codex" },
		now,
	);
	appendRateLimitWindows(
		windows,
		payload.code_review_rate_limit ?? payload.codeReviewRateLimit,
		{ id: "code-review", category: "code_review", name: "Code Review" },
		now,
	);

	const additional = payload.additional_rate_limits ?? payload.additionalRateLimits;
	if (Array.isArray(additional)) {
		for (const [index, item] of additional.entries()) {
			if (!isRecord(item)) continue;
			const name =
				normalizeString(item.limit_name ?? item.limitName) ??
				normalizeString(item.metered_feature ?? item.meteredFeature) ??
				`Additional ${index + 1}`;
			appendRateLimitWindows(
				windows,
				item.rate_limit ?? item.rateLimit,
				{
					id: `additional-${index + 1}-${slug(name)}`,
					category: "additional",
					name,
				},
				now,
			);
		}
	}
	return windows;
}

interface WindowGroup {
	id: string;
	category: CodexQuotaCategory;
	name: string;
}

function appendRateLimitWindows(
	target: CodexQuotaWindow[],
	value: unknown,
	group: WindowGroup,
	now: number,
): void {
	if (!isRecord(value)) return;
	const allowed = booleanValue(value.allowed);
	const limitReached =
		booleanValue(value.limit_reached ?? value.limitReached) === true ||
		allowed === false;
	const candidates = [
		{ source: "primary" as const, value: value.primary_window ?? value.primaryWindow },
		{
			source: "secondary" as const,
			value: value.secondary_window ?? value.secondaryWindow,
		},
	]
		.filter(
			(candidate): candidate is { source: "primary" | "secondary"; value: JsonObject } =>
				isRecord(candidate.value),
		)
		.map((candidate) => ({
			...candidate,
			kind: quotaWindowKind(candidate.value, candidate.source),
		}))
		.sort((left, right) => windowKindRank(left.kind) - windowKindRank(right.kind));

	for (const [index, candidate] of candidates.entries()) {
		const rawUsed = numberValue(
			candidate.value.used_percent ?? candidate.value.usedPercent,
		);
		const usedPercent =
			rawUsed === null
				? limitReached
					? 100
					: null
				: clampPercent(rawUsed);
		const limitWindowSeconds = positiveNumber(
			candidate.value.limit_window_seconds ??
				candidate.value.limitWindowSeconds,
		);
		target.push({
			id: `${group.id}-${candidate.kind}-${index}`,
			category: group.category,
			name: group.name,
			kind: candidate.kind,
			usedPercent,
			remainingPercent:
				usedPercent === null ? null : clampPercent(100 - usedPercent),
			limitWindowSeconds,
			resetAt: quotaResetAt(candidate.value, now),
			allowed,
			limitReached,
		});
	}
}

function quotaWindowKind(
	window: JsonObject,
	fallback: "primary" | "secondary",
): CodexQuotaWindowKind {
	const seconds = positiveNumber(
		window.limit_window_seconds ?? window.limitWindowSeconds,
	);
	if (seconds === FIVE_HOUR_SECONDS) return "five_hour";
	if (seconds === WEEK_SECONDS) return "weekly";
	if (
		seconds !== null &&
		seconds >= MIN_MONTH_SECONDS &&
		seconds <= MAX_MONTH_SECONDS
	) {
		return "monthly";
	}
	return fallback;
}

function quotaResetAt(window: JsonObject, now: number): number | null {
	const absolute = dateValueToMs(window.reset_at ?? window.resetAt);
	if (absolute !== null) return absolute;
	const offset = positiveNumber(
		window.reset_after_seconds ?? window.resetAfterSeconds,
	);
	return offset === null ? null : now + offset * 1000;
}

function normalizeResetCredits(value: unknown): CodexRateLimitResetCredits {
	if (!isRecord(value)) {
		return { availableCount: null, applicableAvailableCount: null };
	}
	return {
		availableCount: nonNegativeNumber(
			value.available_count ?? value.availableCount,
		),
		applicableAvailableCount: nonNegativeNumber(
			value.applicable_available_count ?? value.applicableAvailableCount,
		),
	};
}

function decodeJwt(token: string | undefined): JsonObject | undefined {
	if (!token) return undefined;
	const parts = token.split(".");
	if (parts.length !== 3) return undefined;
	const payload = parts[1];
	if (!payload) return undefined;
	try {
		const normalized = payload.replace(/-/g, "+").replace(/_/g, "/");
		const padded = normalized.padEnd(Math.ceil(normalized.length / 4) * 4, "=");
		const bytes = Uint8Array.from(atob(padded), (character) =>
			character.charCodeAt(0),
		);
		const value: unknown = JSON.parse(new TextDecoder().decode(bytes));
		return isRecord(value) ? value : undefined;
	} catch {
		return undefined;
	}
}

function dateValueToMs(value: unknown): number | null {
	const numeric = numberValue(value);
	if (numeric !== null) {
		if (numeric <= 0) return null;
		const milliseconds = numeric < 1e11 ? numeric * 1000 : numeric;
		return Number.isNaN(new Date(milliseconds).getTime())
			? null
			: milliseconds;
	}
	const text = normalizeString(value);
	if (!text) return null;
	const parsed = Date.parse(text.replace(/(\.\d{6})\d+/, "$1"));
	return Number.isFinite(parsed) ? parsed : null;
}

function normalizePlanType(value: unknown): string | null {
	return normalizeString(value)?.toLowerCase() ?? null;
}

function normalizeString(value: unknown): string | null {
	if (typeof value !== "string") return null;
	const trimmed = value.trim();
	return trimmed || null;
}

function numberValue(value: unknown): number | null {
	if (typeof value === "number") return Number.isFinite(value) ? value : null;
	if (typeof value !== "string" || !value.trim()) return null;
	const parsed = Number(value);
	return Number.isFinite(parsed) ? parsed : null;
}

function positiveNumber(value: unknown): number | null {
	const normalized = numberValue(value);
	return normalized !== null && normalized > 0 ? normalized : null;
}

function nonNegativeNumber(value: unknown): number | null {
	const normalized = numberValue(value);
	return normalized !== null && normalized >= 0 ? normalized : null;
}

function booleanValue(value: unknown): boolean | null {
	return typeof value === "boolean" ? value : null;
}

function clampPercent(value: number): number {
	return Math.max(0, Math.min(100, value));
}

function windowKindRank(kind: CodexQuotaWindowKind): number {
	switch (kind) {
		case "five_hour":
			return 0;
		case "weekly":
			return 1;
		case "monthly":
			return 2;
		case "primary":
			return 3;
		case "secondary":
			return 4;
	}
}

function slug(value: string): string {
	return (
		value
			.trim()
			.toLowerCase()
			.replace(/[^a-z0-9]+/g, "-")
			.replace(/^-+|-+$/g, "") || "quota"
	);
}

async function discardBody(response: Response): Promise<void> {
	try {
		await response.body?.cancel();
	} catch {
		// The safe status-only error remains useful when the body cannot be cancelled.
	}
}

function clientAbortError(error: unknown): DOMException {
	return isAbortError(error)
		? error
		: new DOMException("The client request was aborted.", "AbortError");
}

function codexUsageUnavailable(): ApiError {
	return new ApiError(
		502,
		"Unable to retrieve Codex subscription usage.",
		"upstream_error",
		"codex_usage_unavailable",
	);
}

function codexUsageUpstreamError(status: number): ApiError {
	return new ApiError(
		502,
		"The Codex subscription usage request was rejected.",
		"upstream_error",
		status === 429 ? "codex_usage_rate_limited" : "codex_usage_upstream_error",
	);
}

function invalidCodexUsageResponse(): ApiError {
	return new ApiError(
		502,
		"The Codex subscription usage response was invalid.",
		"upstream_error",
		"invalid_codex_usage_response",
	);
}

File: worker-rs/src/upstream/codex/subscription.rs
use base64::{
    Engine as _,
    engine::general_purpose::{URL_SAFE, URL_SAFE_NO_PAD},
};
use serde::{Deserialize, Serialize};
use serde_json::Value;
use time::{OffsetDateTime, format_description::well_known::Rfc3339};

use crate::core::{ApiError, AppResult, JsonObject, record_field};

pub const CODEX_USAGE_PATH: &str = "/backend-api/wham/usage";
pub const CODEX_USAGE_REQUEST_TIMEOUT_MS: u64 = 10_000;
pub const MAX_CODEX_USAGE_RESPONSE_BYTES: usize = 256 * 1024;
const FIVE_HOUR_SECONDS: f64 = 5.0 * 60.0 * 60.0;
const WEEK_SECONDS: f64 = 7.0 * 24.0 * 60.0 * 60.0;
const MIN_MONTH_SECONDS: f64 = 28.0 * 24.0 * 60.0 * 60.0;
const MAX_MONTH_SECONDS: f64 = 31.0 * 24.0 * 60.0 * 60.0;
const MAX_JAVASCRIPT_DATE_MS: f64 = 8_640_000_000_000_000.0;

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum CodexQuotaWindowKind {
    FiveHour,
    Weekly,
    Monthly,
    Primary,
    Secondary,
}

impl CodexQuotaWindowKind {
    const fn as_str(self) -> &'static str {
        match self {
            Self::FiveHour => "five_hour",
            Self::Weekly => "weekly",
            Self::Monthly => "monthly",
            Self::Primary => "primary",
            Self::Secondary => "secondary",
        }
    }

    const fn rank(self) -> u8 {
        match self {
            Self::FiveHour => 0,
            Self::Weekly => 1,
            Self::Monthly => 2,
            Self::Primary => 3,
            Self::Secondary => 4,
        }
    }
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum CodexQuotaCategory {
    Codex,
    CodeReview,
    Additional,
}

#[derive(Debug, Clone, Default, PartialEq, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct CodexSubscriptionMetadata {
    pub plan_type: Option<String>,
    pub subscription_active_start: Option<f64>,
    pub subscription_active_until: Option<f64>,
}

#[derive(Debug, Clone, PartialEq, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct CodexQuotaWindow {
    pub id: String,
    pub category: CodexQuotaCategory,
    pub name: String,
    pub kind: CodexQuotaWindowKind,
    pub used_percent: Option<f64>,
    pub remaining_percent: Option<f64>,
    pub limit_window_seconds: Option<f64>,
    pub reset_at: Option<f64>,
    pub allowed: Option<bool>,
    pub limit_reached: bool,
}

#[derive(Debug, Clone, Default, PartialEq, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct CodexRateLimitResetCredits {
    pub available_count: Option<f64>,
    pub applicable_available_count: Option<f64>,
}

#[derive(Debug, Clone, PartialEq, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct CodexSubscriptionInfo {
    pub plan_type: Option<String>,
    pub subscription_active_start: Option<f64>,
    pub subscription_active_until: Option<f64>,
    pub windows: Vec<CodexQuotaWindow>,
    pub rate_limit_reset_credits: CodexRateLimitResetCredits,
    pub fetched_at: f64,
}

/// Extracts display metadata from an ID token accepted by the authentication boundary.
pub fn codex_subscription_metadata(id_token: Option<&str>) -> CodexSubscriptionMetadata {
    let claims = id_token.and_then(decode_jwt);
    let auth = record_field(claims.as_ref(), "https://api.openai.com/auth").or(claims.as_ref());
    CodexSubscriptionMetadata {
        plan_type: normalize_plan_type(alias4(
            auth,
            "chatgpt_plan_type",
            "chatgptPlanType",
            "plan_type",
            "planType",
        )),
        subscription_active_start: date_value_to_ms(alias4(
            auth,
            "chatgpt_subscription_active_start",
            "chatgptSubscriptionActiveStart",
            "subscription_active_start",
            "subscriptionActiveStart",
        )),
        subscription_active_until: date_value_to_ms(alias4(
            auth,
            "chatgpt_subscription_active_until",
            "chatgptSubscriptionActiveUntil",
            "subscription_active_until",
            "subscriptionActiveUntil",
        )),
    }
}

pub fn codex_subscription_from_usage(
    payload: &Value,
    metadata: CodexSubscriptionMetadata,
    now_ms: f64,
) -> AppResult<CodexSubscriptionInfo> {
    let payload = payload
        .as_object()
        .ok_or_else(invalid_codex_usage_response)?;
    Ok(CodexSubscriptionInfo {
        plan_type: normalize_plan_type(alias(payload, "plan_type", "planType"))
            .or(metadata.plan_type),
        subscription_active_start: metadata.subscription_active_start,
        subscription_active_until: metadata.subscription_active_until,
        windows: build_quota_windows(payload, now_ms),
        rate_limit_reset_credits: normalize_reset_credits(alias(
            payload,
            "rate_limit_reset_credits",
            "rateLimitResetCredits",
        )),
        fetched_at: now_ms,
    })
}

pub fn codex_usage_unavailable() -> ApiError {
    ApiError::new(502, "Unable to retrieve Codex subscription usage.")
        .with_kind("upstream_error")
        .with_code("codex_usage_unavailable")
}

pub fn codex_usage_upstream_error(status: u16) -> ApiError {
    ApiError::new(502, "The Codex subscription usage request was rejected.")
        .with_kind("upstream_error")
        .with_code(if status == 429 {
            "codex_usage_rate_limited"
        } else {
            "codex_usage_upstream_error"
        })
}

pub fn invalid_codex_usage_response() -> ApiError {
    ApiError::new(502, "The Codex subscription usage response was invalid.")
        .with_kind("upstream_error")
        .with_code("invalid_codex_usage_response")
}

#[derive(Clone, Copy)]
struct WindowGroup<'a> {
    id: &'a str,
    category: CodexQuotaCategory,
    name: &'a str,
}

fn build_quota_windows(payload: &JsonObject, now_ms: f64) -> Vec<CodexQuotaWindow> {
    let mut windows = Vec::new();
    append_rate_limit_windows(
        &mut windows,
        alias(payload, "rate_limit", "rateLimit"),
        WindowGroup {
            id: "codex",
            category: CodexQuotaCategory::Codex,
            name: "Codex",
        },
        now_ms,
    );
    append_rate_limit_windows(
        &mut windows,
        alias(payload, "code_review_rate_limit", "codeReviewRateLimit"),
        WindowGroup {
            id: "code-review",
            category: CodexQuotaCategory::CodeReview,
            name: "Code Review",
        },
        now_ms,
    );
    if let Some(additional) =
        alias(payload, "additional_rate_limits", "additionalRateLimits").and_then(Value::as_array)
    {
        for (index, item) in additional.iter().enumerate() {
            let Some(item) = item.as_object() else {
                continue;
            };
            let name = normalize_string(alias(item, "limit_name", "limitName"))
                .or_else(|| normalize_string(alias(item, "metered_feature", "meteredFeature")))
                .unwrap_or_else(|| format!("Additional {}", index + 1));
            let id = format!("additional-{}-{}", index + 1, slug(&name));
            append_rate_limit_windows(
                &mut windows,
                alias(item, "rate_limit", "rateLimit"),
                WindowGroup {
                    id: &id,
                    category: CodexQuotaCategory::Additional,
                    name: &name,
                },
                now_ms,
            );
        }
    }
    windows
}

fn append_rate_limit_windows(
    target: &mut Vec<CodexQuotaWindow>,
    value: Option<&Value>,
    group: WindowGroup<'_>,
    now_ms: f64,
) {
    let Some(value) = value.and_then(Value::as_object) else {
        return;
    };
    let allowed = value.get("allowed").and_then(Value::as_bool);
    let limit_reached = alias(value, "limit_reached", "limitReached").and_then(Value::as_bool)
        == Some(true)
        || allowed == Some(false);
    let mut candidates = [
        (
            CodexQuotaWindowKind::Primary,
            alias(value, "primary_window", "primaryWindow").and_then(Value::as_object),
        ),
        (
            CodexQuotaWindowKind::Secondary,
            alias(value, "secondary_window", "secondaryWindow").and_then(Value::as_object),
        ),
    ]
    .into_iter()
    .filter_map(|(fallback, window)| {
        window.map(|window| (quota_window_kind(window, fallback), window))
    })
    .collect::<Vec<_>>();
    candidates.sort_by_key(|(kind, _)| kind.rank());
    for (index, (kind, window)) in candidates.into_iter().enumerate() {
        let raw_used = number_value(alias(window, "used_percent", "usedPercent"));
        let used_percent = raw_used
            .map(clamp_percent)
            .or_else(|| limit_reached.then_some(100.0));
        target.push(CodexQuotaWindow {
            id: format!("{}-{}-{index}", group.id, kind.as_str()),
            category: group.category,
            name: group.name.to_owned(),
            kind,
            used_percent,
            remaining_percent: used_percent.map(|value| clamp_percent(100.0 - value)),
            limit_window_seconds: positive_number(alias(
                window,
                "limit_window_seconds",
                "limitWindowSeconds",
            )),
            reset_at: quota_reset_at(window, now_ms),
            allowed,
            limit_reached,
        });
    }
}

fn quota_window_kind(window: &JsonObject, fallback: CodexQuotaWindowKind) -> CodexQuotaWindowKind {
    let seconds = positive_number(alias(window, "limit_window_seconds", "limitWindowSeconds"));
    match seconds {
        Some(value) if value == FIVE_HOUR_SECONDS => CodexQuotaWindowKind::FiveHour,
        Some(value) if value == WEEK_SECONDS => CodexQuotaWindowKind::Weekly,
        Some(value) if (MIN_MONTH_SECONDS..=MAX_MONTH_SECONDS).contains(&value) => {
            CodexQuotaWindowKind::Monthly
        }
        _ => fallback,
    }
}

fn quota_reset_at(window: &JsonObject, now_ms: f64) -> Option<f64> {
    date_value_to_ms(alias(window, "reset_at", "resetAt")).or_else(|| {
        positive_number(alias(window, "reset_after_seconds", "resetAfterSeconds"))
            .map(|seconds| now_ms + seconds * 1_000.0)
    })
}

fn normalize_reset_credits(value: Option<&Value>) -> CodexRateLimitResetCredits {
    let Some(value) = value.and_then(Value::as_object) else {
        return CodexRateLimitResetCredits::default();
    };
    CodexRateLimitResetCredits {
        available_count: nonnegative_number(alias(value, "available_count", "availableCount")),
        applicable_available_count: nonnegative_number(alias(
            value,
            "applicable_available_count",
            "applicableAvailableCount",
        )),
    }
}

fn decode_jwt(token: &str) -> Option<JsonObject> {
    let mut parts = token.split('.');
    let _header = parts.next()?;
    let payload = parts.next()?;
    let _signature = parts.next()?;
    if payload.is_empty() || parts.next().is_some() {
        return None;
    }
    let bytes = URL_SAFE_NO_PAD
        .decode(payload)
        .or_else(|_| URL_SAFE.decode(payload))
        .ok()?;
    serde_json::from_slice::<Value>(&bytes)
        .ok()?
        .as_object()
        .cloned()
}

fn date_value_to_ms(value: Option<&Value>) -> Option<f64> {
    if let Some(numeric) = number_value(value) {
        if numeric <= 0.0 {
            return None;
        }
        let milliseconds = if numeric < 1e11 {
            numeric * 1_000.0
        } else {
            numeric
        };
        return (milliseconds.abs() <= MAX_JAVASCRIPT_DATE_MS).then_some(milliseconds);
    }
    let text = normalize_string(value)?;
    parse_rfc3339_ms(&text)
}

fn parse_rfc3339_ms(value: &str) -> Option<f64> {
    let timestamp = OffsetDateTime::parse(value, &Rfc3339).ok()?;
    let milliseconds = (timestamp.unix_timestamp_nanos() as f64 / 1_000_000.0).trunc();
    (milliseconds.abs() <= MAX_JAVASCRIPT_DATE_MS).then_some(milliseconds)
}

fn normalize_plan_type(value: Option<&Value>) -> Option<String> {
    normalize_string(value).map(|value| value.to_lowercase())
}

fn normalize_string(value: Option<&Value>) -> Option<String> {
    value?
        .as_str()
        .map(str::trim)
        .filter(|value| !value.is_empty())
        .map(str::to_owned)
}

fn number_value(value: Option<&Value>) -> Option<f64> {
    match value? {
        Value::Number(value) => value.as_f64().filter(|value| value.is_finite()),
        Value::String(value) if !value.trim().is_empty() => value
            .trim()
            .parse::<f64>()
            .ok()
            .filter(|value| value.is_finite()),
        _ => None,
    }
}

fn positive_number(value: Option<&Value>) -> Option<f64> {
    number_value(value).filter(|value| *value > 0.0)
}

fn nonnegative_number(value: Option<&Value>) -> Option<f64> {
    number_value(value).filter(|value| *value >= 0.0)
}

fn clamp_percent(value: f64) -> f64 {
    value.clamp(0.0, 100.0)
}

fn slug(value: &str) -> String {
    let mut slug = String::new();
    let mut separator = false;
    for character in value.trim().to_lowercase().chars() {
        if character.is_ascii_alphanumeric() {
            if separator && !slug.is_empty() {
                slug.push('-');
            }
            separator = false;
            slug.push(character);
        } else {
            separator = true;
        }
    }
    if slug.is_empty() {
        "quota".into()
    } else {
        slug
    }
}

fn alias<'a>(object: &'a JsonObject, first: &str, second: &str) -> Option<&'a Value> {
    object
        .get(first)
        .filter(|value| !value.is_null())
        .or_else(|| object.get(second).filter(|value| !value.is_null()))
}

fn alias4<'a>(
    object: Option<&'a JsonObject>,
    first: &str,
    second: &str,
    third: &str,
    fourth: &str,
) -> Option<&'a Value> {
    let object = object?;
    [first, second, third, fourth]
        .into_iter()
        .find_map(|key| object.get(key).filter(|value| !value.is_null()))
}
